contracts/RockPaperScissors.sol

```diff
@@ -51,72 +54,121 @@ contract RockPaperScissors is Pausable {
+        emit LogSubtractedFromBalance(a, amount);
+    }
+
+    function zeroOutGame(bytes32 _hash) private {
+        games[_hash] = Game(msg.sender, Moves.ROCK, false, 0, 0); // leave address populated to not reuse gameHash
+    }
+
+    function startGame(bytes32 _gameHash, uint _wager) public {
+        require(games[_gameHash].b == address(0), "Hash already used");
+        require(_wager <= balances[msg.sender], "Balance too low for this wager");
+        require(_wager * 2 > _wager, "Wager is invalid"); // including when wager == 0 ;)
         
-        games[_gameHash].a = msg.sender;
-        games[_gameHash].wager = msg.value;
-        emit LogStartGame(msg.sender, _gameHash, msg.value);
+        safeSubtractFromBalance(msg.sender, _wager);
+        games[_gameHash].wager = _wager;
+        emit LogStartedGame(msg.sender, _gameHash, _wager);
     }
 
-    function joinGame(bytes32 _gameHash, Moves _move) public payable {
-        require(games[_gameHash].a != address(0), "_gameHash is invalid");
+    function joinGame(bytes32 _gameHash) public {
+        uint wager = games[_gameHash].wager;
+        require(wager > 0, "_gameHash is invalid");
         require(games[_gameHash].b == address(0), "_gameHash already has two players");
-        require(msg.value == games[_gameHash].wager, "msg.value does not match game wager");
-        // TODO: allow players to use balances[msg.sender] to wager
+        require(wager <= balances[msg.sender], "Balance too low for this wager");
         
+        safeSubtractFromBalance(msg.sender, wager);
         games[_gameHash].b = msg.sender;
+        uint deadline = now + secondsUntilForfeit;
+        games[_gameHash].deadline = deadline;
+        emit LogJoinedGame(msg.sender, _gameHash, deadline);
+    }
+
+    function submitMove(bytes32 _gameHash, Moves _move) public {
+        require(games[_gameHash].b == msg.sender, "msg.sender does not match Player B for this game");
+        require(!games[_gameHash].bMoveSubmitted, "Move already submitted for this game");
+
         games[_gameHash].bMove = _move;
-        games[_gameHash].deadline = now + secondsUntilForfeit;
-        emit LogJoinGame(msg.sender, _gameHash, _move, games[_gameHash].deadline);
+        games[_gameHash].bMoveSubmitted = true;
+        uint deadline = now + secondsUntilForfeit;
+        games[_gameHash].deadline = deadline;
+        
+        emit LogSubmittedMove(msg.sender, _gameHash, _move, deadline);
     }
 
     function endGame(bytes32 _salt, Moves _aMove) public {
         bytes32 gameHash = createHash(_salt, msg.sender, _aMove);
-        require(games[gameHash].a == msg.sender, "You didn't start this game");
-        require(!games[gameHash].completed, "Game already completed");
+        uint wager = games[gameHash].wager;
+        require(games[gameHash].bMoveSubmitted, "Player B hasn't submitted move yet");
+        require(games[gameHash].wager > 0, "Game doesn't exist or already completed");
```

Comment:
This `require` is probably redundant; I don't think there's a game state where `bMoveSubmitted` (previous line) is true and `wager == 0`.

Reply:
I think you're right.